src/engine/effects/engineeffectsdelay.cpp

```diff
@@ -0,0 +1,152 @@
+#include "engine/effects/engineeffectsdelay.h"
+
+#include "util/sample.h"
+
+void EngineEffectsDelay::setDelay(unsigned int delaySamples) {
+    m_delaySamples = delaySamples;
+}
+
+void EngineEffectsDelay::process(const CSAMPLE* pIn, CSAMPLE* pOutput, const int iBufferSize) {
+    if (m_oldDelaySamples == 0 && m_delaySamples == 0) {
+        if (pIn != pOutput) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+        }
+
+        return;
+    }
+
+    if (m_oldDelaySamples == m_delaySamples &&
+            m_delayBufferSize == iBufferSize + m_delaySamples) {
+        int delaySourcePos = (m_delayPos + m_delayBufferSize - m_delaySamples) % m_delayBufferSize;
+
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos >= 0) {
+            if (pIn != pOutput) {
+                SampleUtil::copy(pOutput, pIn, iBufferSize);
+            }
+
+            return;
+        }
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos <= static_cast<int>(m_delayBufferSize)) {
+            if (pIn != pOutput) {
+                SampleUtil::copy(pOutput, pIn, iBufferSize);
+            }
+
+            return;
+        }
+
+        for (int i = 0; i < iBufferSize; ++i) {
+            // Put samples into delay buffer.
+            m_delayBuffer[m_delayPos] = pIn[i];
+            m_delayPos = (m_delayPos + 1) % m_delayBufferSize;
+
+            // Take a delayed sample from the delay buffer and copy it to the destination buffer.
+            pOutput[i] = m_delayBuffer[delaySourcePos];
+            delaySourcePos = (delaySourcePos + 1) % m_delayBufferSize;
+        }
+
+    } else {
+        // Size of the new delay buffer.
+        int newDelayBufferSize = iBufferSize + m_delaySamples;
+
+        if (m_delayBufferSize == 0) {
+            // Creates first delay buffer and sets all elements to zero value.
+
+            m_delayBuffer = new CSAMPLE[newDelayBufferSize];
```

Comment:
New is locking and must not be called from the engine.
Just allocated a reasonable maximum during the constructor.

Reply:
Okay, thank you. It will be a better option. I will fix it. The buffer size wasn't allocated in the constructor due to the huge amount of delay of the Pitch shift effect, but it will be solved.